### PR TITLE
2019-09-22 - Beta test for network boot

### PR DESCRIPTION
**Pull request by timg236**
@XECDesign please could you create a .deb file on the internal server ? It should be fine but I'll double check first + update the Pi4s on the CI servers.

### CODE REVIEW

**Comment by kohenkatz** [comment]
Is this at a point where it can/should be tested by regular users who are interested in beta testing it, or should we wait?

File: firmware/raspberry_pi4_network_boot_beta.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+## Server configuration
+Network boot requires a TFTP and NFS server to be configured.  See [Network boot server tutorial](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bootmodes/net_tutorial.md)
+
+Additional notes:-
+* The VC firmware (start4.elf) must be updated to the latest rpi-update version in order to get the Pi4 network drivers.
+* The MAC address on the Pi4 is programmed at manufacture and is not derived from the serial number.
+
+```
+# mac adddress (ip addr) - it should start with dc:a6:3
```
Comment (andrum99) [inline firmware/raspberry_pi4_network_boot_beta.md]: Unless you're got another OUI in the pipeline that the IEEE list (http://standards-oui.ieee.org/oui.txt) doesn't know about, that should have a 2 on the end, i.e. dc:a6:32

File: firmware/raspberry_pi4_network_boot_beta.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+## Installation - Updating firmware and bootloader
+```
+# Install the rpi-eeprom update package
+sudo apt update
+sudo apt upgrade
+sudo apt install rpi-eeprom
+
+# Get the latest VC firmware which includes the Pi4 network driver in start.elf
```
Comment (andrum99) [inline firmware/raspberry_pi4_network_boot_beta.md]: start.elf -> start4.elf ?

File: firmware/raspberry_pi4_network_boot_beta.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+Network boot is not enabled by default in the bootloader. To enable it the bootloader configuration file must be edited.
+```
+# Extract the configuration file
+cp /lib/firmware/raspberrypi/beta/pieeprom-2019-09-22.bin pieeprom.bin
+rpi-eeprom-config pieeprom.bin > bootconf.txt
+
+# Enable network boot
+# change boot_order from 0x1 (sd-boot) to 
+BOOT_ORDER=0x21
+# sd-boot then failover to network boot
```
Comment (andrum99) [inline firmware/raspberry_pi4_network_boot_beta.md]: Presumably BOOT_ORDER=0x20 would give us network boot only, without SD boot?

File: firmware/raspberry_pi4_network_boot_beta.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+### Update the bootloader EEPROM
+```
+# Flash the bootloader EEPROM with network boot enabled
+# Run 'rpi-eeprom-update -h' for more information
+sudo rpi-eeprom-update -d -f ./pieeprom-netboot.bin
+sudo reboot
+```
+
+## TFTP prefix
+As with the Pi3 bootloader the presence of /serial_number/start.elf is used to detect
```
Comment (andrum99) [inline firmware/raspberry_pi4_network_boot_beta.md]: Again, should this be start4.elf on the Pi 4?

File: firmware/raspberry_pi4_network_boot_beta.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+The bootloader loads start4.elf in preference to start.elf so it's possible to
+just have an empty start.elf file if you only want to have Pi4 firmware in the TFTP
+root directory.
+
+## Bootloader configuration
+
+### BOOT_ORDER
+The BOOT_ORDER setting allows flexible configuration for the priority of different
+bootmodes. It is represented as 32bit unsigned integer where each nybble represents
```
Comment (andrum99) [inline firmware/raspberry_pi4_network_boot_beta.md]: nybble -> nibble

File: firmware/raspberry_pi4_network_boot_beta.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+**This is an early beta test and has only been tested on a small number of networks. Check with your network administrator before enabling network boot.**
+
+This documentation will be merged into the main bootloader documents after the beta test is complete.
+
+## Server configuration
+Network boot requires a TFTP and NFS server to be configured.  See [Network boot server tutorial](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bootmodes/net_tutorial.md)
+
+Additional notes:-
+* The VC firmware (start4.elf) must be updated to the latest rpi-update version in order to get the Pi4 network drivers.
```
Comment (andrum99) [inline firmware/raspberry_pi4_network_boot_beta.md]: VC -> VideoCore ?

File: firmware/raspberry_pi4_network_boot_beta.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+## TFTP prefix
+As with the Pi3 bootloader the presence of /serial_number/start.elf is used to detect
+whether the serial number should be used as the directory prefix for TFTP requests.
+
+The bootloader loads start4.elf in preference to start.elf so it's possible to
+just have an empty start.elf file if you only want to have Pi4 firmware in the TFTP
+root directory.
+
```
Comment (andrum99) [inline firmware/raspberry_pi4_network_boot_beta.md]: start4.elf is only used on the Pi 4 - is that right? And the Pi 4 doesn't use start.elf at all, except for this check for a serial number-specific directory on the TFTP server - is that right?

File: firmware/raspberry_pi4_network_boot_beta.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+Default: 4000  
+Minimum: 500  
+
+### TFTP_TIMEOUT
+The timeout in milliseconds for an individual file download via TFTP.  
+Default: 15000  
+Minimum: 5000  
+
+### TFTP_IP
+Optional dotted decimal ip address for the TFTP server which overrides the server-ip from the DHCP request.
```
Comment (andrum99) [inline firmware/raspberry_pi4_network_boot_beta.md]: ip -> IP

Also, might be useful to give an example of what a dotted decimal IP address is - presumably that's one like 192.168.1.250 ?

**Comment by andrum99** [comment]
Sorry about the bugspam - I will add my suggested minor fixes to raspberry_pi4_network_boot_beta.md as a PR once this integrates.

**Comment by timg236** [comment]
Fixed some minor issues in bootconf + some doc changes. replacing this with 2019-09-23

**Comment by timg236** [comment]
> Is this at a point where it can/should be tested by regular users who are interested in beta testing it, or should we wait?

It's probably best to wait until the Debian package is pushed and an we post something on the forums. However, feel to try the EEPROM packages before then and post comments. The worst case is that you might have use the bootloader rescue image and perhaps with network boot there's a small risk that it could generate garbage on the network

**Comment by timg236** [comment]
@andrum99 Thanks for reviewing the PR, I think I managed to incorporate most of them when I updated the PR for the 2019-09-23 which fixed a bootconf.txt issue. Happy to accept PRs for any new issues on the main branch now that it's all merged.

**Pull request closed without merging**